### PR TITLE
[Feature:InstructorUI] Polls data dump

### PR DESCRIPTION
**Pull request by BenKelly99**
### What is the current behavior?
Currently, there is no way to access poll data without sysadmin access.

### What is the new behavior?
Now, any instructor can get access to a JSON summary of all the polls, so they can use it to assign late days, give extra points, etc... The JSON is generated in the .../reports/ directory when "Generate Grade Summaries" is run. The file is only generated if polling is enabled.

### CODE REVIEW

**Comment by MasterOdin** [comment]
For a course like CS1 or CS2 (or any course with a lot of students), how big is this json file going to get?

**Comment by bmcutler** [comment]
For Data Structures: 28 lectures x 6 multiple choice questions per lecture x # of students x 700 students. 
For Fall 2020 (with ~300 students) a dump (different format) of this data is ~700kb.

**Comment by BenKelly99** [comment]
@MasterOdin @bmcutler I added 28 * 6 = 168 polls to the database, each with a title of ~30 characters and then added 700 student responses to each of them. The file ended up being 3.3M.

**Comment by MasterOdin** [comment]
@bmcutler Thanks for providing that information

@BenKelly99 Thanks for running those numbers to see the load. I am somewhat cautious on this part of the codebase on eating up too much memory, but I think we can spare 3.3M overhead in this singular function and can leave this function simpler above complicating it with streaming out the JSON file piece by piece as we do elsewhere in the report generator.

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:89]:
```suggestion
            if (!is_writable($poll_base_path)) {
```

**Comment by BenKelly99** [comment]
So leave as is for now? Sounds good.

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:86]: I would be in favor of moving this before we generate the grade summaries themselves. That way, if we have `is_writable` failures, we hit those first before attempting to write out any data. My general preference on error handling is to cover as much up-front (especially these sort of static checks) so that we don't have to give return messages to the user like "Generated grade summaries, but failed to generate poll summaries" or whatever.

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:397]: Very slight micro optimization, but we can save on the memory steps of creating a new variable and assigning data to it and instead just push the entire thing to `$polls_data` once we're ready:

```suggestion
            $student_response_data = [];
            foreach ($poll->getUserResponses() as $student => $response) {
                $student_response_data[$student] = $poll->isCorrect($response);
            }
            $poll_data["responses"] = $student_response_data;
            $polls_data[] = [
                'name' => $poll->getName(),
                'responses' => $student_response_data,
            ];
```

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:386]:
```suggestion
    private function generatePollSummaryInternal(string $base_path): void {
```

**Review by MasterOdin — APPROVED** [review]
Slight comment, but otherwise the code looks and works fine.

For reference, this looks to add 3 queries per poll to the generation page. Not exactly a huge deal here (I think) with ~90 new queries, but something that can probably be revised in the future to take advantage of joins and such to reduce that number. Same with looking at lowering the memory. Could be lowered, but not important I think at this stage as <3.5M is not that much memory in the scheme of things for a page that is not that often run.

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:395]:
```suggestion
````

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:73]: I was suggestion moving the entire if block here, and not just the error check, but this is fine. I do think we still want to run this error check only if polls are enabled or else in an incorrectly configured setup, someone might hit this error with polls disabled, and then complain about this as a bug that polls are doing something they shouldn't be.

```suggestion
        if ($this->core->getConfig()->isPollsEnabled() && !is_writable($poll_base_path)) {
```

**Comment by BenKelly99** [comment]
Combined this PR with the import/export PR since they became closely related after the changes Barb and I discussed.

File: site/app/controllers/PollController.php
Comment (MasterOdin) [inline site/app/controllers/PollController.php:433]: Put all elements on their own lines. This will (eventually) be a lint rule:

```suggestion
                !array_key_exists("name", $poll)
                || !array_key_exists("question", $poll)
                || !array_key_exists("responses", $poll)
                || !array_key_exists("correct_responses", $poll)
                || !array_key_exists("release_date", $poll)
```

File: site/app/controllers/PollController.php
Comment (MasterOdin) [inline site/app/controllers/PollController.php:454]:
```suggestion
        if ($num_errors === 0) {
```

File: site/app/controllers/admin/ReportController.php
Comment (MasterOdin) [inline site/app/controllers/admin/ReportController.php:398]: Use [`FileUtils::writeJsonFile`](https://github.com/Submitty/Submitty/blob/a9e09405eaf6b370fdd8978805a144a71042306e/site/app/libraries/FileUtils.php#L287):

```suggestion
        FileUtils::writeJsonFile(FileUtils::joinPaths($base_path, "poll_responses_summary.json"), $polls_data);
        FileUtils::writeJsonFile(FileUtils::joinPaths($base_path, "poll_data_summary.json"), PollUtils::getPollExportData($polls));
```

File: site/app/libraries/PollUtils.php
Comment (MasterOdin) [inline site/app/libraries/PollUtils.php:7]: unused `use` statement:

```suggestion
```

File: site/app/libraries/PollUtils.php
Comment (MasterOdin) [inline site/app/libraries/PollUtils.php:19]: can you add a test case or two (e.g. test for empty `$polls` array, test for array with few elements) for this method?

**Review by MasterOdin — CHANGES_REQUESTED** [review]

File: site/app/controllers/PollController.php
Comment (MasterOdin) [reply site/app/controllers/PollController.php:406]: This isn't asking expressly to change anything, just for clarification for myself to understand this code / purpose.

Comment (BenKelly99) [reply site/app/controllers/PollController.php:406]: It should just be for the ajax download. If there's a better way to make this function to reflect that, let me know and I can change it.

Comment (MasterOdin) [reply site/app/controllers/PollController.php:407]: You could mimic some of the functions in [MiscController](https://github.com/Submitty/Submitty/blob/master/site/app/controllers/MiscController.php) (like displayFile) where you turn off header/footer, manually set some headers to register it as a download, and then use `$this->core->getOutput()->renderString` to output the JSON. The JS could then be removed and replaced with a regular `<a>` tag. If you want to do it, that's fine, can also hold off and I think first put in place a generic `DownloadResponse` that makes doing this easier around the system.

Let me know which you'd prefer to do here.

File: migration/migrator/migrations/course/20210111163413_add_poll_data_folder.py
Comment (MasterOdin) [inline migration/migrator/migrations/course/20210111163413_add_poll_data_folder.py:19]: We want to use built-in python stuff as much as possible so that we get error handling and such if things blow up. By using `os.system`, we would shallow any possible errors that might happen there. So for chown, we can use [`shutil.chown`](https://docs.python.org/3/library/shutil.html#shutil.chown). As this then relies on `os.chown`, we can also use the `Path` directly without having to first convert it to a string.

```suggestion
    shutil.chown(polls_dir, php_user, course_group_id)
```

**Review by MasterOdin — APPROVED** [review]

File: site/app/controllers/PollController.php
Comment (BenKelly99) [reply site/app/controllers/PollController.php:407]: I'll work on making a DownloadResponse. Seems like it will be useful.

Comment (BenKelly99) [reply site/app/controllers/PollController.php:407]: For the DownloadResponse, would it make sense for it to just take in a path to a file that will be downloaded? Then it can grab the mime info from the file and download it that way.

Comment (MasterOdin) [reply site/app/controllers/PollController.php:407]: Let's not do the DownloadResponse within this PR, and instead do it separately (after this is merged in the next day or so) as then that separate PR can address all of the places such a new class would benefit, and the class can be built to cater to all of them from the get-go. For doing that, can you spin this conversation into a new issue and then also include all of the functions that would benefit from this new DownloadResponse? This should then give us a good idea of where we would expect to utilize this to users. I expect that some of the methods would utilize a file (and so would use the php function `readfile` as well) while others are just going to output some string buffer (e.g. this method).

Comment (BenKelly99) [reply site/app/controllers/PollController.php:407]: Roger that. I've updated this PR to use the same method as MiscController::displayFile which should be easy to change over once the new Response type is made.

**Comment by bmcutler** [comment]
@MasterOdin @BenKelly99 
Is this in a state to be merged & installed?

Also what is the timing to remove the feature flag gating on polling?
Instructors would like to use this for Spring 2021 term :)

**Comment by BenKelly99** [comment]
@bmcutler I think it should be ready to go unless Matt has comments on the new HTTP header version of the download. I can remove the feature flag in this PR or open a separate PR for that purpose. Let me know which you would prefer.

**Comment by bmcutler** [comment]
Please open another PR to remote the feature flag

**Pull request merged**
